compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java

```diff
@@ -1592,45 +1598,120 @@ private BObjectType getDistinctObjectType(BLangTypeDefinition typeDefinition, BO
         return definedObjType;
     }
 
-    private BType getCyclicDefinedType(BLangTypeDefinition typeDef, SymbolEnv env) {
-        BUnionType unionType = BUnionType.create(null, new LinkedHashSet<>());
-        unionType.isCyclic = true;
-        Name typeDefName = names.fromIdNode(typeDef.name);
-
-        BTypeSymbol typeDefSymbol = Symbols.createTypeSymbol(SymTag.UNION_TYPE, Flags.asMask(typeDef.flagSet),
-                typeDefName, env.enclPkg.symbol.pkgID, unionType, env.scope.owner,
-                typeDef.name.pos, SOURCE);
-
-        typeDef.symbol = typeDefSymbol;
-        unionType.tsymbol = typeDefSymbol;
-
-        // We define the unionType in the main scope here
+    private void defineTypeInMainScope(BTypeSymbol typeDefSymbol, BLangTypeDefinition typeDef) {
         if (PackageID.isLangLibPackageID(this.env.enclPkg.packageID)) {
             typeDefSymbol.origin = BUILTIN;
             handleLangLibTypes(typeDef);
         } else {
             defineSymbol(typeDef.name.pos, typeDefSymbol);
         }
+    }
+
+    private BType defineSymbolsForCyclicTypeDefinitions(BLangTypeDefinition typeDef, SymbolEnv env) {
+        Name newTypeDefName = names.fromIdNode(typeDef.name);
+        BTypeSymbol typeDefSymbol;
+        BType newTypeNode;
+
+        switch (typeDef.typeNode.getKind()) {
+            case TUPLE_TYPE_NODE:
+                newTypeNode = new BTupleType(null, new ArrayList<>());
+                typeDefSymbol = Symbols.createTypeSymbol(SymTag.TUPLE_TYPE, Flags.asMask(typeDef.flagSet),
+                        newTypeDefName, env.enclPkg.symbol.pkgID, newTypeNode, env.scope.owner,
+                        typeDef.name.pos, SOURCE);
+                break;
+            default:
+                newTypeNode = BUnionType.create(null, new LinkedHashSet<>());
+                typeDefSymbol = Symbols.createTypeSymbol(SymTag.UNION_TYPE, Flags.asMask(typeDef.flagSet),
+                        newTypeDefName, env.enclPkg.symbol.pkgID, newTypeNode, env.scope.owner,
+                        typeDef.name.pos, SOURCE);
+        }
+        newTypeNode.isCyclic = true;
+        BSymbol foundSym = symResolver.lookupSymbolInMainSpace(env, names.fromIdNode(typeDef.name));
```

Comment:
This was to prevent defining the symbol in main scope again in the case of a cyclic reference in a scenario like `Foo[int, Foo...]`

Reply:
You have to define by calling `defineTypeInMainScope` for this `Foo` type to be available in the scope right.
So don't we should call `defineTypeInMainScope` first then try to resolve the type.

Reply:
I'm checking in the scope before defining to support scenarios like : 
```
type XType1 XNil|XTuple1|XNever|XAny|();
const XNil = "nil";
const XNever = "never";
const XAny = "any";
type XTuple1 ["tuple", XType1, XType1];
```
Since we will be defining `XTuple1` when defining `XType1 type` when evaluating `XTuple1 type`, if we define in the scope first before checking then it will be an issue right.

Reply:
I agreed

Reply:
Reconsider the logic. I think better to define the types before hand. Check this below code and how to handle it.
```Ballerina
type XType1 XNil|XTuple1[]|XNever|XAny|();
const XNil = "nil";
const XNever = "never";
const XAny = "any";
type XTuple1 ["tuple", XType1, XType1];
```
```LogTalk
 [2021-06-16 09:05:46,264] SEVERE {b7a.log.crash} - class org.wso2.ballerinalang.compiler.semantics.model.types.BNoType cannot be cast to class org.wso2.ballerinalang.compiler.semantics.model.types.BUnionType (org.wso2.ballerinalang.compiler.semantics.model.types.BNoType and org.wso2.ballerinalang.compiler.semantics.model.types.BUnionType are in unnamed module of loader 'app') 
java.lang.ClassCastException: class org.wso2.ballerinalang.compiler.semantics.model.types.BNoType cannot be cast to class org.wso2.ballerinalang.compiler.semantics.model.types.BUnionType (org.wso2.ballerinalang.compiler.semantics.model.types.BNoType and org.wso2.ballerinalang.compiler.semantics.model.types.BUnionType are in unnamed module of loader 'app')
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.getCyclicDefinedType(SymbolEnter.java:1701)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SymbolEnter.visit(SymbolEnter.java:1367)

```

See  https://github.com/ballerina-platform/ballerina-lang/pull/30683/files#r652277608 also